File: examples/go/docs/public_orderbook_service_get_orderbook.go
// Copyright 2019 Tulip Solutions B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"context"
	"fmt"
	"github.com/tulipsolutions/tecl/tulipsolutions/api/common/orders"
	"github.com/tulipsolutions/tecl/tulipsolutions/api/pub/orderbook"
	"google.golang.org/grpc"
	"os"
	"time"
)

func publicOrderbookServiceGetOrderbook(conn *grpc.ClientConn, parentContext context.Context) {
	client := orderbook.NewPublicOrderbookServiceClient(conn)

	// Create a request for the BTC_EUR orderbook, with the greatest precision and largest length
	// See TBD for semantics of Precision and Length
	request := orderbook.GetOrderbookRequest{
		Market:    orders.Market_BTC_EUR,
		Precision: orderbook.Precision_P3,
		Length:    orderbook.Length_NUM_ENTRIES_100,
	}

	// Create a new context with a 1s deadline and make the request synchronously
	ctx, cancel := context.WithTimeout(parentContext, time.Second)
	defer cancel()
	response, err := client.GetOrderbook(ctx, &request)
	if err != nil {
		_, _ = fmt.Fprintf(os.Stderr, "PublicOrderbookService.GetOrderbook error: %v \n", err)
		return
	}
	fmt.Println(response)
}

File: examples/python/docs/public_orderbook_service_get_orderbook.py
# Copyright 2019 Tulip Solutions B.V.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import print_function

import sys

import grpc

from tulipsolutions.api.common import orders_pb2
from tulipsolutions.api.pub import orderbook_pb2, orderbook_pb2_grpc


def public_orderbook_service_get_orderbook(channel):
    # CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
    stub = orderbook_pb2_grpc.PublicOrderbookServiceStub(channel)

    # Create a request for the BTC_EUR orderbook, with the greatest precision, largest length,
    # and highest update frequency
    # See TBD for semantics of Precision and Length
    request = orderbook_pb2.GetOrderbookRequest(
        market=orders_pb2.BTC_EUR,
        precision=orderbook_pb2.P0,
        length=orderbook_pb2.NUM_ENTRIES_25,
    )

    try:
        # Make the request synchronously with a 1s deadline
        response = stub.GetOrderbook(request, timeout=1)
        print(response)
    except grpc.RpcError as e:
        print("PublicOrderbookService.GetOrderbook error: " + str(e), file=sys.stderr)
    # CODEINCLUDE-END-MARKER: ref-code-example-request

    # CODEINCLUDE-BEGIN-MARKER: ref-code-example-response
    result_string = "{}\n".format(type(response).__name__)
    for orderbook_entry in response.entries:
        result_string += "\t{} {} {} orders @ {} total {}\n".format(
            type(orderbook_entry).__name__,
            orders_pb2.Side.Name(orderbook_entry.side),
            orderbook_entry.orders_at_price_level,
            orderbook_entry.price_level,
            orderbook_entry.amount,
        )
    print(result_string)
    # CODEINCLUDE-END-MARKER: ref-code-example-response
